File: 16-mcp-model-context-protocol/16-4-building-custom-mcp-serve.py
# mcp_server.py
from mcp.server import Server
from mcp.server.stdio import stdio_server

server = Server("my-python-server")

@server.tool()
async def my_tool(param1: str, param2: int = 0) -> str:
    """Does something useful with the provided input.
    
    Args:
        param1: First parameter (required)
        param2: Optional number parameter
    """
    result = await process_input(param1, param2)
    return json.dumps(result)

async def main():
    async with stdio_server() as (read_stream, write_stream):
        await server.run(read_stream, write_stream)

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())

File: 16-mcp-model-context-protocol/16-4-building-custom-mcp-serve.js
// mcp-server.js
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';

const server = new Server({
  name: 'my-custom-server',
  version: '1.0.0'
});

// Define available tools
server.setRequestHandler('tools/list', async () => ({
  tools: [{
    name: 'my_tool',
    description: 'Does something useful with the provided input',
    inputSchema: {
      type: 'object',
      properties: {
        param1: { 
          type: 'string', 
          description: 'First parameter' 
        },
        param2: {
          type: 'number',
          description: 'Optional number parameter'
        }
      },
      required: ['param1']
    }
  }]
}));

// Handle tool calls
server.setRequestHandler('tools/call', async (request) => {
  if (request.params.name === 'my_tool') {
    const { param1, param2 } = request.params.arguments;
    const result = await processInput(param1, param2);
    return { 
      content: [{ 
        type: 'text', 
        text: JSON.stringify(result, null, 2) 
      }] 
    };
  }
  throw new Error(`Unknown tool: ${request.params.name}`);
});

// Start server
const transport = new StdioServerTransport();
await server.connect(transport);